### DISCUSSION TITLE
Multiple User Sessions

### DISCUSSION THREAD

**Discussion by kstoddard** [remix-run/remix#4498] [Q&A] [answered]
It appears to me that if a user session is set up per the docs or patterned after a common stack (e.g., Blues Stack), that multiple user sessions will collide. If you look at the source code for sessions (below) you can see that Remix grabs the value of the session cookie to query session storage for the data. In the docs and in the Blues Stack nothing unique about the user gets tied to the session cookie. This means that only one user can use the system at a time without issue. I've have managed to negate this issue by following this process instead...

```jsx
export async function action({ request }) {
  const response = await api.login(input);

  // some validation checks for a user object here

  // This sets the value of the `sessionCookie` to equal the user's id
  // `sessionCookie` is being created/exported same way as most examples (with sessionStorage setup)
  const cookie = await sessionCookie.serialize(`userId-${response.user.id}`);

  // NOW get the session based on the unique id just set on the `sessionCookie`
  const session = await getSession(cookie);

  // Add data to the session
  session.set('user', {
    id: user.id,
    token: user.token,
  });

  return redirect('/clients', {
    headers: {
      'Set-Cookie': await commitSession(session)
    },
  })
}
```

I am wondering if anyone indeed knows that this is an issue and if not point to an example where multiple user sessions are being used per the docs. I have set things up just like the Blues Stack and confirmed that multiple user sessions get clobbered with the latest authenticated user. I have also verified the above snippet and multiple user sessions work.

The strategy in play in the above example is...
- Authenticate the user
- Import the session cookie (same cookie used to set up session for the app)
- Add a unique value to the session cookie based on user info
- THEN pass the session cookie to the `getSession` helper
- Everything from the last point forward works exactly like the docs and other examples show

I have also brought this up in Discord but there has been no movement (see: https://discord.com/channels/770287896669978684/1037426454130065428/1037426454130065428). You can check out that thread if you are logged in and have access to the Remix Discord.

**Reference:**

- https://remix.run/docs/en/v1/api/remix#using-sessions
- https://github.com/remix-run/blues-stack/blob/main/app/session.server.ts#L78
- https://github.com/remix-run/remix/blob/main/packages/remix-server-runtime/sessions.ts#L238

**Comment by sergiodxa** [comment, marked as answer]
The way the session data is stored depends on the session storage mechanism.

The most commonly used one is the cookie session storage where the session data is stored directly in the cookie.

In this scenario, each browser will contain a cookie with whatever you store there that lets you identify the user, usually this is either a token or the userId.

In other session storages the cookie contains a sessionId which is used to find the session data in another data store like an in memory object, the file system or a data base.

In all cases the cookie name don’t need to change and users will never access another user session data, but if you don’t store anything to identify the user in the session there’s no way to know who us that user.

**Reply by kstoddard** [reply]
@sergiodxa you are correct but in my testing the only scenario that actually works as described is when the session storage object is the cookie itself, since of course, it only ever lives on a users machine. When using a storage mechanism that persists somehow (e.g., file, db, Redis, etc.) merely adding data to the session falls down. Look closely at the source code I linked to, when `getSession` is called it expects a unique identifier to be stored on the cookie which it in turn passes off to your session backend. In all examples I have found this unique identifier remains empty supporting exactly one user at a time. 

In the example code I provide, a unique identifier gets added to the session cookie prior to getting the session (after auth), which correctly separates each session on the backend. 

I am building out user auth in a system now at work and when following the docs and authenticating as different users from different browsers, the latest session always overrides the previous. Again, in studying the source I found that a unique identifier is required to correctly separate each session and no one is doing this anywhere. My code above works though 🔥 🤷🏼‍♂️

**Reply by kiliman** [reply]
Hmm.. I'n not sure why you're seeing that behavior. Here's a Codesandbox sample that uses file-based session storage (data stored in fileystem, session cookie only contains session id).

Open 2 browser instances (one incognito) and you'll see that the session data is different for each.

https://codesandbox.io/s/remix-file-sessions-s1yf3f

**Reply by kstoddard** [reply]
@kiliman I will do a little more digging and follow up with a example that can be reproduced. Thanks for sharing! I may need to also investigate the differences (if any) between the session drivers.

**Reply by kstoddard** [reply]
@sergiodxa @kiliman Alas I can no longer reproduce the issue which is good but alarming because I spent a healthy amount of time testing this before submitting the discussion 😬. In any case, sessions are correctly being separated regardless of the driver I choose at this point. We are in development and have been using the `createMemorySessionStorage` until we set up Redis and I thought perhaps maybe the issue was _just_ with this driver but no, can't reproduce with the problem with it now either. Thanks again for your help, the Remix community has been fabulous to work with!

**Reply by sergiodxa** [reply]
Note that `createMemorySessionStorage` doesn't work in development out of the box, because Remix clears the require cache on every request to ensure the server code is up to date, but this cause memory objects to be cleared on every request, the way to make it work is to store it on `global.something`.
